### DISCUSSION TITLE
[General] ProviderError: execution reverted when sending ETH to payable method

### DISCUSSION THREAD

**Discussion by waelsy123** [zkSync-Community-Hub/zksync-developers#315] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I'm getting execution reverted when sending ETH to payable method 

### What is the specific issue or error you're encountering?

  const tx = await contract.connect(signer).payWithEtherAndRequestQuery(query, queryType,
        { value: ethers.parseEther("0.001") }
    );
    
    I got ProviderError: execution reverted  error when passing the last param:    { value: ethers.parseEther("0.001") }
 

### Can you share the error messages or logs you're receiving, if any?

   
    ProviderError: execution reverted
    at HttpProvider.request (/greeter-example/node_modules/hardhat/src/internal/core/providers/http.ts:88:21)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async HardhatEthersProvider.estimateGas (/greeter-example/node_modules/@nomicfoundation/hardhat-ethers/src/internal/hardhat-ethers-provider.ts:237:27) 

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?
```
 function payWithEtherAndRequestQuery(
        string calldata sqlQuery,
        string calldata queryType
    ) external payable {
        payable(relayerContract).transfer(msg.value);
        (bool success, ) = relayerContract.call(
            abi.encodeWithSignature("query(string,string)", sqlQuery, queryType)
        );
        require(success, "Query request failed");
    }
```
### Have you tried to isolate the problem, and if so, what were the results?

yes, it works if I remove the native payment

### What steps have you already taken to try to resolve the issue?

I tried the same with Ethereum and it works there

**Comment by githubdoramon** [comment]
@waelsy123 seems like you are using .transfer without passing the gas amount. This is discouraged on Ethereum itself, and not supported on Era. Read more here - https://docs.zksync.io/build/quick-start/best-practices.html#use-call-over-send-or-transfer


Let me know if this works.

**Comment by waelsy123** [comment]
I believe the issue is mainly connected to the client building the transaction. usually if I want to send ETH along when calling a method I pass it as last arg, ie: 
```
const tx = await contract.connect(signer).somelogic(arg1,
{ value: ethers.parseEther("0.1") }
);
```

as you see above `{ value: ethers.parseEther("0.001") }` should send 0.1 ETH that will appear in msg.value. that should normally work but here it cause an error.

**Reply by danijelTxFusion** [reply, marked as answer]
What is the `signer` in your case? I think the signer that is provided is not capable to form valid transaction. Here's the simple [contract](https://sepolia.explorer.zksync.io/address/0xF47ABe1801E522C4Dbd878dc51849646462bAF92#contract)  that has payable method and the script how to interact with it using `zksync-ethers: 6.0.0` :

```ts
import { ethers } from "ethers";
import { Provider, types, Contract, Wallet } from "zksync-ethers";

const PRIVATE_KEY = process.env.PRIVATE_KEY as string;
const CONTRACT_ADDRESS = "0xF47ABe1801E522C4Dbd878dc51849646462bAF92";
const CONTRACT_ABI = `[
  {
    "inputs": [
      {
        "internalType": "string",
        "name": "_greeting",
        "type": "string"
      }
    ],
    "stateMutability": "nonpayable",
    "type": "constructor"
  },
  {
    "inputs": [],
    "name": "greet",
    "outputs": [
      {
        "internalType": "string",
        "name": "",
        "type": "string"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "string",
        "name": "_greeting",
        "type": "string"
      }
    ],
    "name": "setGreeting",
    "outputs": [],
    "stateMutability": "payable",
    "type": "function"
  }
]`


async function main() {
  const provider = Provider.getDefaultProvider(types.Network.Sepolia);
  const wallet = new Wallet(PRIVATE_KEY, provider);

  const contract = new Contract(
    CONTRACT_ADDRESS,
    CONTRACT_ABI,
    wallet
  );

  console.log(`Balance berofe tx: ${await provider.getBalance(await contract.getAddress())}`);

  const transaction = await contract.setGreeting("Hello people!", {value: ethers.parseEther("0.001")});
  console.log(`Transaction hash of setting new message: ${transaction.hash}`);

  await transaction.wait();

  console.log(`The message now is: ${await contract.greet()}`);
  console.log(`Balance after tx: ${await provider.getBalance(await contract.getAddress())}`);

}


main()
    .then()
    .catch((error) => {
        console.log(`Error: ${error}`);
    });
```

Output:

```
Balance berofe tx: 0
Transaction hash of setting new message: 0x90f0f0bf9340988f940ac659882fb90b3dd2d90a3a52645ef74be52260def755
The message now is: Hello people!
Balance after tx: 1000000000000000
Done in 3.44s.
```

**Comment by loriyamazaki** [comment]
It seems like the issue you're encountering is related to sending Ether to a payable function in your contract on the Testnet environment. The error message "ProviderError: execution reverted" suggests that the transaction execution failed and reverted within the Ethereum Virtual Machine (EVM).

Here are a few potential reasons for this issue and steps you can take to resolve it:

Gas Limit: Ensure that you're providing enough gas for the transaction to be executed successfully. If the gas limit is too low, the transaction may run out of gas and revert. Try increasing the gas limit provided for the transaction.

Contract Logic: Review the logic within your payWithEtherAndRequestQuery function. Ensure that it properly handles the Ether sent to it and that there are no issues with the relayerContract or the query function call. Double-check that the function is designed to handle incoming Ether correctly.

Contract State: Check if there are any state variables or contract conditions that could cause the transaction to revert unexpectedly. Make sure that the contract is in the expected state when executing the transaction.

Testnet Specifics: Testnets may have different conditions or limitations compared to the Ethereum mainnet. Verify that the Testnet you're using supports the features and functionalities required by your contract, including native payments.

Debugging: Consider adding additional logging or debugging statements within your contract function to get more insights into where the transaction is failing. This can help pinpoint the exact cause of the issue.

By addressing these potential areas of concern and thoroughly debugging your contract code, you should be able to identify and resolve the issue with sending Ether to the payable function on the Testnet environment.

**Comment by mganlok** [comment]
The error "ProviderError: execution reverted" indicates that the transaction execution failed and reverted. In your Solidity function payWithEtherAndRequestQuery, ensure that relayerContract is a valid address and that the query function is implemented correctly in the contract. Also, check if the function is properly handling the received Ether. Double-check your contract logic and consider debugging further to identify the root cause of the revert. #Solidity #TransactionRevert #PayableFunction

**Comment by sonhoranemichelle** [comment]
Ensure that the `relayerContract` address is correctly set and that the contract's `query` function is properly implemented to handle the provided parameters.

**Comment by mapeshewa** [comment]
The error "ProviderError: execution reverted" is occurring when passing the { value: ethers.parseEther("0.001") } parameter when calling the payWithEtherAndRequestQuery() function in the contract. The function works without the native payment. It works on Ethereum but not on Testnet.
